Vtt support for transcripts

Shaac opened the pull request:
### Description

This PR adds VTT support for transcripts (in addition to existing SRT and JSON).

Closes #7431 

### Checklist
<!-- 
  To help us keep the issue tracker clean and work as efficient as possible,
  please make sure that you have done all of the following.
  You can tick the boxes below by placing an x inside the brackets like this: [x]
-->
- [x] I have read the contribution guidelines: https://github.com/AntennaPod/AntennaPod/blob/develop/CONTRIBUTING.md#submit-a-pull-request
- [x] I have performed a self-review of my code
- [x] I have run the automated code checks using `./gradlew checkstyle spotbugsPlayDebug spotbugsDebug :app:lintPlayDebug`
- [x] My code follows the style guidelines of the AntennaPod project: https://github.com/AntennaPod/AntennaPod/wiki/Code-style
- [x] I have mentioned the corresponding issue and the relevant keyword (e.g., "Closes: #xy") in the description (see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] If it is a core feature, I have added automated tests

ByteHamster commented:
Replying to the comment in the issue here: https://github.com/AntennaPod/AntennaPod/issues/7431#issuecomment-2512865782

> They are definitely some functions that should be factorize, like parseTimecode

Would be great if you could do that :)

> I should probably write a test VTT file with all the features and either make a fake podcast with it or add some integration tests with it.

Having a test file would be great, yeah. There are already some tests for other formats here: https://github.com/AntennaPod/AntennaPod/blob/develop/parser/transcript/src/test/java/de/danoeh/antennapod/parser/transcript/ Would be great if you could follow the format (or, if the file is large, add it to the assets directory)

ByteHamster commented:
Hi @Shaac are you still interested in this issue? Would be great if we could get vtt support merged :)

Shaac commented:
> Hi @Shaac are you still interested in this issue? Would be great if we could get vtt support merged :)

Sorry, I let myself be distracted by other tasks.
I’ll put this higher on my priority list.

I’m still very interested in having vtt support indeed!
I’ll try to make some progress very soon.

Shaac commented:
I’ve updated the PR with a state that I now consider ready to be merged (after review of course).

I changed the approach here: instead of doing like we do for SRT, a parsing of several blocks, then making our our blocks, keeping in memory who is the speaker through different blocks, I do something way simpler.

The way VTT is used, both in practice, and in the [podcast 2.0 specs](https://github.com/Podcastindex-org/podcast-namespace/blob/main/transcripts/transcripts.md), we have longer lines, and the speaker in specified at each block. So no need for complex reprocessing of the input: we can keep each block as its own, and it comes directly with its own speaker.

This allows for more simple and straight forward code than the existing SRT parser. This also means there is no longer any copy paste here. The only function that could be factorized is the very simple `VttTranscriptParser::parseTimestamp` function, that is very similar to `SrtTranscriptParser::parseTimecode`, but still different (it’s not the same regex, and in VTT the hours can be omitted). So I could make a util function with this method, taking into input the regex, but the function is so small and so simple that in my opinion it would do more harm than good. But if you still think it’s better, I can make a util class (but I’m not sure where: a `transcript/TranscriptUtils.java` class?)

ByteHamster commented:
I cleaned up the type detection a bit to avoid having to specify the mime types in several different locations.

> The way VTT is used, both in practice, and in the [podcast 2.0 specs](https://github.com/Podcastindex-org/podcast-namespace/blob/main/transcripts/transcripts.md), we have longer lines

The lines are still quite short. At least in the `podnews.net/rss` feed, the lines fill only just a bit more than one line of the screen in AntennaPod, making the transcript page look a bit jagged. I think adding together several very short lines (like in the other transcript types) could help to make it look nicer.

> the function is so small and so simple that in my opinion it would do more harm than good.

I agree

Shaac commented:
> I cleaned up the type detection a bit to avoid having to specify the mime types in several different locations.

Great it’s better thanks!

 
> > The way VTT is used, both in practice, and in the [podcast 2.0 specs](https://github.com/Podcastindex-org/podcast-namespace/blob/main/transcripts/transcripts.md), we have longer lines
> 
> The lines are still quite short. At least in the `podnews.net/rss` feed, the lines fill only just a bit more than one line of the screen in AntennaPod, making the transcript page look a bit jagged. I think adding together several very short lines (like in the other transcript types) could help to make it look nicer.

So I looked at a bunch a vtt from podnews (like [this one](https://podnews.net/audio/podnews250219.mp3.vtt) without speakers and [this one](https://podnews.net/audio/podnews250213.mp3.vtt) with speaker tags), and reread the [current podcast 2.0 specification](https://github.com/Podcastindex-org/podcast-namespace/blob/main/transcripts/transcripts.md#webvtt), and a few things strike me:

- some lines are short indeed (and the specification even recommends 65 characters per line), and it looks like on podnews they have an automatic cut if their line is bigger than 65 characters
- they don’t even (both the spec and podnews) use proper VTT specification for author: from what I understood from the spec, the author tag should be repeated at each line

I tried to look at other podcasts that use VTT, but I did not really found some significant one that could be used as an other example. I looked at the transcript generated by Apple (that AntennaPod will never see since they are for Apple Podcast app only), and they have very long lines (one line per sentence), each with a proper speaker tag.

So I suppose people use something like whisper to generate their subtitles, but some will cut the sentences at 65 characters, and some will keep very long lines. And then in AntennaPod we’d want to guess what they meant, and undo the splitting at 65 that was done before… It’s a bit of a mess.

There should probably a standard to specify if the lines were automatically cut and should be grouped together again. A way to hint that some postprocess should be done on the client side. Anyway.

What I propose I do is:

- If there is a speaker tag, consider all subsequent lines with be from that same speaker (unless a line starts with a new speaker). I don’t think it’s proper VTT, but it is what seems to be in usage for podcasting
- Do the same strategy as for SRT: if a transcript line takes less than 5 seconds, then the next line will be merged with it (if from the same speaker). Several short lines can be merged that way

But for the last point, I’m wondering if we should only do it if there is no author before. Maybe it’s a way of distinguishing VTT that are automatically split, and the rest.

```
00:00:05.000 --> 00:00:06.000
<v John>This line has a tag,

00:00:06.000 --> 00:00:07.000
while this one has not.

00:00:07.000 --> 00:00:08.000
<v John> Still John talking.
```

In the example above, we’d have the first two line joined, but not the third.

Maybe it allows us to detect what was the intent of the person who wrote the VTT. But maybe I’m just wrong and trying to invent a new standard inside the standard.

ByteHamster commented:
> they have very long lines (one line per sentence)

I wouldn't call that long, to be honest. One sentence only fills something like one line on my screen. I would aim for about 5 lines per "segment".

> If there is a speaker tag, consider all subsequent lines with be from that same speaker

Sounds good 👍 

> Do the same strategy as for SRT: if a transcript line takes less than 5 seconds, then the next line will be merged with it (if from the same speaker). Several short lines can be merged that way

Sounds good as well. Now, I wonder if we should make this two components. One parsing the file and one going over the transcript again and doing the merging. That way, we avoid code duplication between the different transcript types and can probably significantly simplify the parser implementation. However, this then prevents format-specific "guessing" like the one you describe:

> Maybe it allows us to detect what was the intent of the person who wrote the VTT.

I haven't looked at too many transcripts yet. Does this approach generate segments of about 5 seconds? Maybe we could somehow use it as an indicator - it could be a good point to break, but still merge if it is too short otherwise? This would be an argument against having a global format-agnostic "merging" code

keunes commented:
I don't know if it complicates matters, but it from a UX/UI perspective I would:

* Collate everything of a speaker if there are speakers, collate up to X lines if there are not.
* Do highlighting still per initial block, to maintain the highest level of detail possible.

It would then look something like this, for one speaker block (just a mock-up using my notes app):
![Imagepipe_0.jpg](https://github.com/user-attachments/assets/296b427f-2114-4c25-943b-8ab962b88a2a)

ByteHamster commented:
I think sub-block highlighting is currently out of scope for this PR. That would introduce a lot of UI changes instead of just adding the parser. For now, I would create multiple blocks that are each not too long but also not too short

Shaac commented:
I’ve given all of it some thoughts over time, and the more I think about it, the more I think this is too complex to try and do something clever.
I’ll update the PR next week with grouping lines together until they reach 5 seconds, like we do for the other formats.
That what we keep things simple, and we can merge it. Maybe we’ll think of improvements in the future.

ByteHamster commented:
Hi @Shaac, are you still interested in this issue? :)

Shaac commented:
Hello,
Sorry I got caught life by life again.

I just did the modifications I said I would:

- if a speaker is not specified, keep the same speaker as before
- group together short lines until it makes for a block of 5 seconds (and not more than 8 seconds)

Note that for group duration, I did not use the same method as used for SRT and JSON: I took the time between the last word and the first one. It seemed to make more sense than summing the duration of each line (we don’t want to group two lines if there is a big silence between them – although I doubt in practice there will be much silences in a podcast).

I updated the unit tests to reflect these two changes. I somehow could not test with the emulator, somehow android studio does not launch the new compiled app; I’ll have to investigate more tomorrow. But apart from not being able to test in the emulator yet, it seems to me that the feature is now completed.

Shaac commented:
Follow-up: I did a clean rebuild to have the change working on the emulator (don’t know what messed up my worflow) and was able to test it: it looks like it working as we intended!

ByteHamster commented on parser/transcript/src/main/java/de/danoeh/antennapod/parser/transcript/VttTranscriptParser.java:
```diff
@@ -0,0 +1,145 @@
+            // should we merge this segment with the previous one?
+            if (segment != null && segment.getSpeaker().equals(speaker)
+                    && timings.end - segment.getStartTime() < TranscriptParser.MAX_SPAN) {
+                segment = new TranscriptSegment(
+                        segment.getStartTime(), timings.end, segment.getWords() + " " + payload, speaker);
```
This creates and then immediately drops quite a lot of segments. How about adding an `append(endTime, words)` method to the transcript segment that can add additional text? That could then maybe also be used to make Srt parsing a bit easier to read (instead of having a bunch of variables)

When first reading this, I was a bit confused why it creates a new segment object. An "append" method would be a bit easier to read for future developers as well

Shaac replied:
It seems a good idea indeed.
I did not want to modify an external class at first, but indeed it would make things easier to read.
I won’t be able to access my dev setup next week but I’ll be able to this after.

Shaac replied:
I added this method as requested

ByteHamster reviewed (commented):
Thanks a lot for working on it, looks very good already :) I have a small suggestion that would probably make the code more easy to read

ByteHamster commented:
Thanks! Will be released in AntennaPod 3.9.0. In case you didn't know yet, we are having a monthly call to meet up with other AntennaPod volunteers. You are warmly invited in case you want to join: https://antennapod.org/events/community-meeting :)
